docs/runtime_image.md

```diff
@@ -0,0 +1,83 @@
+# Runtime image
+
+## Overview
+
+For a dynamic languages like PHP, Python or Ruby, the build-time and run-time environments are the same. In this case using builder as a base image for a resulting image is natural.
+
+For compiled languages like C/C++, Java or Golang, the dependencies necessary for compilation might dramatically outweigh the actual runtime artifacts. To keep runtime images slim, S2I enables multiple-step build processes, where a binary artifact such as an executable or Java WAR file is created in the first builder image, extracted, and injected into an second image that simply places the executable in the correct location for execution. To give you even more abilities to customize a resulting image, S2I is also executing `assemble-runtime` script inside of this (runtime) image. In this way you may do a final adjustments by modifying files before an image will be committed.
+
+The following scheme illustrates the build workflow:
+
+![s2i workflow](./runtime-image-flow.png "s2i workflow")
+
+
+## How it works?
+
+To make this work S2I needs to know the following:
+
+* builder image
+* mapping of the artifacts
+* runtime image
+
+And all these information you should specify during S2I invocation:
+
+    s2i build <repo> <builder-image> <my-app> --runtime-image <runtime-image> --runtime-artifact </path/to/artifact>
+
+The only new here are `--runtime-image` and `--runtime-artifact` options. The latter option is used to specify a full path to file (or directory) inside of the builder image and a corresponding directory on the runtime image where S2I will copy these files. Note that you can specify this option multiple times.
```

Comment:
>because right now you're referring to a "corresponding directory" that you aren't actually specifying in that example, so it's confusing.

Ah, yes, now I see what do you mean. Fixed: https://github.com/openshift/source-to-image/pull/536/commits/f38a5b1310885a96da172aa32c76170a18355204

>I don't want to have to override the assemble-runtime script just because my maven build produces a file named "foo.war" and my assemble-runtime script only looks for "ROOT.war" or something. (There are probably better examples).

Yes, but this option doesn't help in provided case because it only affects destination directory and not file name. In other words, you only can choose whether foo.war will be uploaded as `WORKDIR/foo.war` or `WORKDIR/you-app/foo.war`.

>also it makes that argument consistent with what we already do for injections.

Yes, now I agree. I also realized this point later.